diff --git a/eng/native/configuretools.cmake b/eng/native/configuretools.cmake
index ad5dc38107c33f..860e46427b6b08 100644
--- a/eng/native/configuretools.cmake
+++ b/eng/native/configuretools.cmake
@@ -7,15 +7,15 @@ if (CMAKE_C_COMPILER MATCHES "-?[0-9]+(\.[0-9]+)?$")
 endif()
 
 if(NOT WIN32 AND NOT CLR_CMAKE_TARGET_BROWSER)
-  if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  if(CMAKE_C_COMPILER_ID MATCHES "Clang")
     if(APPLE)
       set(TOOLSET_PREFIX "")
     else()
       set(TOOLSET_PREFIX "llvm-")
     endif()
-  elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+  elseif(CMAKE_C_COMPILER_ID MATCHES "GNU")
     if(CMAKE_CROSSCOMPILING)
-      set(TOOLSET_PREFIX "${CMAKE_CXX_COMPILER_TARGET}-")
+      set(TOOLSET_PREFIX "${CMAKE_C_COMPILER_TARGET}-")
     else()
       set(TOOLSET_PREFIX "")
     endif()
diff --git a/src/mono/mono.proj b/src/mono/mono.proj
index bff71d03b47e9c..ed1945437a6689 100644
--- a/src/mono/mono.proj
+++ b/src/mono/mono.proj
@@ -530,7 +530,7 @@
       <_Objcopy Condition="'$(Platform)' == 's390x'">s390x-linux-$(_LinuxAbi)-$(_Objcopy)</_Objcopy>
       <_Objcopy Condition="'$(Platform)' == 'x64'">x86_64-linux-$(_LinuxAbi)-$(_Objcopy)</_Objcopy>
       <_Objcopy Condition="'$(Platform)' == 'x86'">i686-linux-$(_LinuxAbi)-$(_Objcopy)</_Objcopy>
-      <_Objcopy Condition="'$(TargetsAndroid)' == 'true'">$(ANDROID_NDK_ROOT)/toolchains/llvm/prebuilt/$(MonoToolchainPrebuiltOS)/bin/$(_Objcopy)</_Objcopy>
+      <_Objcopy Condition="'$(TargetsAndroid)' == 'true'">$(ANDROID_NDK_ROOT)/toolchains/llvm/prebuilt/$(MonoToolchainPrebuiltOS)/bin/llvm-objcopy</_Objcopy>
     </PropertyGroup>
     <!-- test viability of objcopy command -->
     <Exec Condition="'$(BuildMonoAOTCrossCompilerOnly)' != 'true' and ('$(TargetsLinux)' == 'true' or '$(TargetsAndroid)' == 'true')" Command="$(_Objcopy) -V" IgnoreStandardErrorWarningFormat="true" ContinueOnError="WarnAndContinue" IgnoreExitCode="true" EchoOff="true" ConsoleToMsBuild="true">
