diff --git a/eng/native/configuretools.cmake b/eng/native/configuretools.cmake
index 37f3b4932cd228..136cd67925d0cb 100644
--- a/eng/native/configuretools.cmake
+++ b/eng/native/configuretools.cmake
@@ -65,12 +65,14 @@ endif()
 
 if (NOT CLR_CMAKE_HOST_WIN32)
   # detect linker
-  set(ldVersion ${CMAKE_C_COMPILER};-Wl,--version)
+  separate_arguments(ldVersion UNIX_COMMAND "${CMAKE_C_COMPILER} ${CMAKE_SHARED_LINKER_FLAGS} -Wl,--version")
   execute_process(COMMAND ${ldVersion}
     ERROR_QUIET
     OUTPUT_VARIABLE ldVersionOutput)
 
-  if("${ldVersionOutput}" MATCHES "GNU ld" OR "${ldVersionOutput}" MATCHES "GNU gold" OR "${ldVersionOutput}" MATCHES "GNU linkers")
+  if("${ldVersionOutput}" MATCHES "LLD")
+    set(LD_LLVM 1)
+  elseif("${ldVersionOutput}" MATCHES "GNU ld" OR "${ldVersionOutput}" MATCHES "GNU gold" OR "${ldVersionOutput}" MATCHES "GNU linkers")
     set(LD_GNU 1)
   elseif("${ldVersionOutput}" MATCHES "Solaris Link")
     set(LD_SOLARIS 1)
